### PR TITLE
feature: added SJF scheduler and Regression based prediction

### PR DESCRIPTION
**Pull request by emil916**
- added SJF scheduler 
- added Regression based prediction, which includes tenant preprocessing logic that executes within a "mini" sandbox
- added logging mechanism for the "mini" sandboxes
- separated the old ADMISSIONS control logic from the execution histogram into two different modules
- upgraded LLVM_VERSION from 12 to 13

### CODE REVIEW

File: runtime/include/module.h [hunk 1]
```diff
@@ -124,7 +121,7 @@ module_initialize_pools(struct module *module)
 static inline void
 module_deinitialize_pools(struct module *module)
 {
-	for (int i = 0; i < runtime_worker_threads_count; i++) {
+	for (int i = 0; i < (module->type == APP_MODULE ? runtime_worker_threads_count : 1); i++) {
```
[the remark this answers is not in the document]
Comment (gparmer) [reply runtime/include/module.h:124]: This is an optimization pass where the compiler tries to "lift" such conditionals out of the loop, but I'm not convinced the compiler will be able to do that here. Given that, I'm essentially suggesting to "manulally lift" the condition out of the loop.

That said, I just realized that this is not on the critical path, so you can ignore all of this advice ;-)

File: runtime/include/sandbox_set_as_error.h [hunk 2]
```diff
@@ -48,14 +49,20 @@ sandbox_set_as_error(struct sandbox *sandbox, sandbox_state_t last_state)
 	/* State Change Bookkeeping */
 	assert(now > sandbox->timestamp_of.last_state_change);
 	sandbox->last_state_duration = now - sandbox->timestamp_of.last_state_change;
+	if (last_state == SANDBOX_RUNNING_SYS)
```
Comment (gparmer) [inline runtime/include/sandbox_set_as_error.h:52]: If this is the style of sledge, then great!

But I'd never want to use multi-line conditions without explicit `{}`. It opens you up to too many accidental control-flow mistakes.

**Review by gparmer — APPROVED** [review]
Amazed how simple you were able to keep this! Great job.

File: runtime/include/execution_regression.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+get_regression_prediction(struct http_session *session)
+{
+	/* Default Pre-processing - Extract payload size */
+	const int payload_size = session->http_request.body_length;
+
+	const double regression_params[2] = { payload_size, session->param2 };
+
+	/* Regression */
+	const struct regression_model model      = session->route->regr_model;
+	const uint64_t                prediction = (regression_params[0] / model.scale * model.beta1
```
[the remark this answers is not in the document]
Comment (emil916) [reply runtime/include/execution_regression.h:18]: Done

File: runtime/include/http_session.h [hunk 4]
```diff
@@ -57,6 +57,9 @@ struct http_session {
 	uint64_t                request_downloaded_timestamp;
 	uint64_t                response_takeoff_timestamp;
 	uint64_t                response_sent_timestamp;
+	bool                    did_preprocessing;
+	uint64_t                preprocessing_duration;
+	double                  param2;
```
[the remark this answers is not in the document]
Comment (emil916) [reply runtime/include/http_session.h:62]: Done

File: runtime/include/module.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (emil916) [reply runtime/include/module.h:124]: Great find! Thank you! Solved.

File: runtime/include/sandbox_set_as_error.h (hunk 2 above)
Comment (emil916) [reply runtime/include/sandbox_set_as_error.h:52]: Done.

**Pull request merged**